### DISCUSSION TITLE
Unit Test Errors in MathComponent Tutorial

### DISCUSSION THREAD

**Discussion by uri-herrera** [nasa/fprime#1135] [Q&A]
Hello,

I'm just about getting used to the F Prime framework however, unit testing with the MathComponent tutorial leads to various warnings and errors upon building. This is after changing the Tester files. Here's the output:

(fprime-venv) uri@uri-VirtualBox:~/Desktop/fprime/Ref/MathSender$ fprime-util build --ut
[WARNING] Failed to find settings file: /home/uri/Desktop/fprime/Ref/settings.ini
[ 14%] Built target codegen
[ 14%] Built target Fw_Cfg
[ 28%] Built target Fw_Types
[ 42%] Built target Fw_Logger
[ 42%] Built target Fw_Obj
[ 42%] Built target Fw_Port
[ 42%] Built target Ref_MathPorts
[ 42%] Built target Fw_Com
[ 57%] Built target Fw_Cmd
[ 71%] Built target Fw_Time
[ 71%] Built target Fw_Log
[ 71%] Built target Fw_Tlm
[ 71%] Built target Utils_Hash
[ 71%] Built target Fw_Comp
[100%] Built target Os
[100%] Built target Fw_CompQueued
[100%] Built target Ref_MathSender
[100%] Built target gtest
[100%] Built target gtest_main
[100%] Linking CXX executable ../../bin/Linux/Ref_MathSender_ut_exe
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/TestMain.o: in function `Nominal_SubOperationTest_Test::TestBody()':
/home/uri/Desktop/fprime/Ref/MathSender/test/ut/TestMain.cpp:10: undefined reference to `Ref::Tester::testSubCommand()'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/TestMain.o: in function `Nominal_MultOperationTest_Test::TestBody()':
/home/uri/Desktop/fprime/Ref/MathSender/test/ut/TestMain.cpp:15: undefined reference to `Ref::Tester::testMultCommand()'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/TestMain.o: in function `Nominal_DivideOperationTest_Test::TestBody()':
/home/uri/Desktop/fprime/Ref/MathSender/test/ut/TestMain.cpp:20: undefined reference to `Ref::Tester::testDivCommand()'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o: in function `Ref::Tester::Tester()':
/home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:29: undefined reference to `Ref::MathSenderGTestBase::MathSenderGTestBase(char const*, unsigned int)'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:29: undefined reference to `Ref::MathSenderGTestBase::~MathSenderGTestBase()'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o: in function `Ref::Tester::~Tester()':
/home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:40: undefined reference to `Ref::MathSenderGTestBase::~MathSenderGTestBase()'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o: in function `Ref::Tester::testAddCommand()':
/home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:52: undefined reference to `Ref::MathSenderTesterBase::sendCmd_MS_DO_MATH(int, unsigned int, float, float, Ref::MathSenderComponentBase::MathOp)'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:56: undefined reference to `Ref::MathSenderGTestBase::assertFromPortHistory_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:58: undefined reference to `Ref::MathSenderGTestBase::assert_from_mathOut_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:62: undefined reference to `Ref::MathSenderGTestBase::assertTlm_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:64: undefined reference to `Ref::MathSenderGTestBase::assertTlm_MS_VAL1_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:65: undefined reference to `Ref::MathSenderGTestBase::assertTlm_MS_VAL2_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:66: undefined reference to `Ref::MathSenderGTestBase::assertTlm_MS_OP_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:68: undefined reference to `Ref::MathSenderGTestBase::assertTlm_MS_VAL1(char const*, unsigned int, unsigned int, float const&) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:69: undefined reference to `Ref::MathSenderGTestBase::assertTlm_MS_VAL2(char const*, unsigned int, unsigned int, float const&) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:70: undefined reference to `Ref::MathSenderGTestBase::assertTlm_MS_OP(char const*, unsigned int, unsigned int, Ref::MathSenderComponentBase::MathOpTlm const&) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:72: undefined reference to `Ref::MathSenderGTestBase::assertEvents_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:74: undefined reference to `Ref::MathSenderGTestBase::assertEvents_MS_COMMAND_RECV_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:76: undefined reference to `Ref::MathSenderGTestBase::assertEvents_MS_COMMAND_RECV(char const*, unsigned int, unsigned int, float, float, Ref::MathSenderComponentBase::MathOpEv) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:78: undefined reference to `Ref::MathSenderGTestBase::assertCmdResponse_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:80: undefined reference to `Ref::MathSenderGTestBase::assertCmdResponse(char const*, unsigned int, unsigned int, unsigned int, unsigned int, Fw::CommandResponse) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:83: undefined reference to `Ref::MathSenderTesterBase::clearHistory()'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:85: undefined reference to `Ref::MathSenderTesterBase::invoke_to_mathIn(int, float)'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:89: undefined reference to `Ref::MathSenderGTestBase::assertTlm_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:91: undefined reference to `Ref::MathSenderGTestBase::assertTlm_MS_RES_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:93: undefined reference to `Ref::MathSenderGTestBase::assertTlm_MS_RES(char const*, unsigned int, unsigned int, float const&) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:95: undefined reference to `Ref::MathSenderGTestBase::assertEvents_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:97: undefined reference to `Ref::MathSenderGTestBase::assertEvents_MS_RESULT_size(char const*, unsigned int, unsigned int) const'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:99: undefined reference to `Ref::MathSenderGTestBase::assertEvents_MS_RESULT(char const*, unsigned int, unsigned int, float) const'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o: in function `Ref::Tester::from_mathOut_handler(int, float, float, Ref::MathOperation)':
/home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:119: undefined reference to `Ref::MathSenderTesterBase::pushFromPortEntry_mathOut(float, float, Ref::MathOperation)'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o: in function `Ref::Tester::connectPorts()':
/home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:131: undefined reference to `Ref::MathSenderTesterBase::connect_to_mathIn(int, Ref::InputMathResultPort*)'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:137: undefined reference to `Ref::MathSenderTesterBase::connect_to_CmdDisp(int, Fw::InputCmdPort*)'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:143: undefined reference to `Ref::MathSenderTesterBase::get_from_mathOut(int)'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:149: undefined reference to `Ref::MathSenderTesterBase::get_from_CmdStatus(int)'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:155: undefined reference to `Ref::MathSenderTesterBase::get_from_CmdReg(int)'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:161: undefined reference to `Ref::MathSenderTesterBase::get_from_Tlm(int)'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:167: undefined reference to `Ref::MathSenderTesterBase::get_from_Time(int)'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:173: undefined reference to `Ref::MathSenderTesterBase::get_from_Log(int)'
/usr/bin/ld: /home/uri/Desktop/fprime/Ref/MathSender/test/ut/Tester.cpp:179: undefined reference to `Ref::MathSenderTesterBase::get_from_LogText(int)'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o:(.data.rel.ro._ZTVN3Ref6TesterE[_ZTVN3Ref6TesterE]+0x28): undefined reference to `Ref::MathSenderTesterBase::init(int)'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o:(.data.rel.ro._ZTVN3Ref6TesterE[_ZTVN3Ref6TesterE]+0x38): undefined reference to `Ref::MathSenderTesterBase::cmdResponseIn(unsigned int, unsigned int, Fw::CommandResponse)'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o:(.data.rel.ro._ZTVN3Ref6TesterE[_ZTVN3Ref6TesterE]+0x40): undefined reference to `Ref::MathSenderTesterBase::textLogIn(unsigned int, Fw::Time&, Fw::TextLogSeverity, Fw::TextLogString const&)'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o:(.data.rel.ro._ZTVN3Ref6TesterE[_ZTVN3Ref6TesterE]+0x48): undefined reference to `Ref::MathSenderTesterBase::logIn_ACTIVITY_LO_MS_COMMAND_RECV(float, float, Ref::MathSenderComponentBase::MathOpEv)'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o:(.data.rel.ro._ZTVN3Ref6TesterE[_ZTVN3Ref6TesterE]+0x50): undefined reference to `Ref::MathSenderTesterBase::logIn_ACTIVITY_HI_MS_RESULT(float)'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o:(.data.rel.ro._ZTVN3Ref6TesterE[_ZTVN3Ref6TesterE]+0x58): undefined reference to `Ref::MathSenderTesterBase::tlmInput_MS_VAL1(Fw::Time const&, float const&)'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o:(.data.rel.ro._ZTVN3Ref6TesterE[_ZTVN3Ref6TesterE]+0x60): undefined reference to `Ref::MathSenderTesterBase::tlmInput_MS_VAL2(Fw::Time const&, float const&)'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o:(.data.rel.ro._ZTVN3Ref6TesterE[_ZTVN3Ref6TesterE]+0x68): undefined reference to `Ref::MathSenderTesterBase::tlmInput_MS_OP(Fw::Time const&, Ref::MathSenderComponentBase::MathOpTlm const&)'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o:(.data.rel.ro._ZTVN3Ref6TesterE[_ZTVN3Ref6TesterE]+0x70): undefined reference to `Ref::MathSenderTesterBase::tlmInput_MS_RES(Fw::Time const&, float const&)'
/usr/bin/ld: CMakeFiles/Ref_MathSender_ut_exe.dir/test/ut/Tester.o:(.data.rel.ro._ZTIN3Ref6TesterE[_ZTIN3Ref6TesterE]+0x10): undefined reference to `typeinfo for Ref::MathSenderGTestBase'
collect2: error: ld returned 1 exit status
make[3]: *** [Ref/MathSender/CMakeFiles/Ref_MathSender_ut_exe.dir/build.make:117: bin/Linux/Ref_MathSender_ut_exe] Error 1
make[2]: *** [CMakeFiles/Makefile2:27374: Ref/MathSender/CMakeFiles/Ref_MathSender_ut_exe.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:27381: Ref/MathSender/CMakeFiles/Ref_MathSender_ut_exe.dir/rule] Error 2
make: *** [Makefile:9912: Ref_MathSender_ut_exe] Error 2
[ERROR] CMake erred with return code 2

I'll run through the entire tutorial again but in the meantime I'd like to know where I went wrong.

Thank you for the help.

**Comment by ThibFrgsGmz** [comment]
Have you been able to make it work since then?

Did you take the latest devel push, the v2.1.0 tag from F' or an older one?

I noticed that you used VirtualBox, did you try with WSL-2 to see?

**Comment by LeStarch** [comment]
It looks like either: 

1. The test commands aren't implemented correctly
2. The Tester.cpp was not included in the UT_SOURCE_FILES list

**Comment by danjwait** [comment]
FWIW, based on the first line in Uri's output I've run into what I think is a similar issue relating to a settings.ini file:

```
djwait@TRON:~/02_Projects/fprime/Ref/MathSender$ mkdir -p test/ut
djwait@TRON:~/02_Projects/fprime/Ref/MathSender$ fprime-util impl --ut
[WARNING] Failed to find settings file: /home/djwait/02_Projects/fprime/Ref/settings.ini
make: *** No rule to make target 'Ref_MathSender_testimpl'.  Stop.
[CMAKE] CMake failed to detect target, attempting CMake cache refresh and retry
make: *** No rule to make target 'Ref_MathSender_testimpl'.  Stop.
[ERROR] CMake erred with return code 2
```
The file is there:
```
djwait@TRON:~/02_Projects/fprime/Ref$ ls *.ini
gds.ini  setting.ini
```
but it's pretty barren:
```
; Ref requires no specific settings thus the [fprime] configuration block is empty
; For more information: https://nasa.github.io/fprime/UsersGuide/user/settings.html
[fprime]
```
I also noticed that in the [4.5.1. Set Up the Unit Test Environment](https://nasa.github.io/fprime/Tutorials/MathComponent/Tutorial.html#The-MathSender-Component_Write-and-Run-Unit-Tests_Set-Up-the-Unit-Test-Environment) section that the order is:
```
mkdir
fprime-util impl --ut
```
but a few lines later it says to run `fprime-util generate --ut` then `fprime-util build --ut' ; could it be an order of operations thing?

I forked devel branch a few days ago, CHANGELOG.md says it's v3.0.0. I'm running on WSL2, Win 11, using VS code editor.

**Reply by astroesteban** [reply]
@danjwait You're correct that steps are in the wrong order now and some steps are missing. I have a pull request out that fixes that issue. You should follow the markdown file here: https://github.com/nasa/fprime/pull/1243

**Reply by danjwait** [reply]
@astroesteban Awesome, thank you! Fixed that, on to Write and Run One Test

**Reply by astroesteban** [reply]
Happy to help 🚀
